amethyst_ui/src/glyphs.rs

```diff
@@ -310,19 +311,54 @@ impl<'a, B: Backend> System<'a> for UiGlyphsSystem<B> {
                     text,
                 };
 
-                ui_text.cached_glyphs.extend(
-                    glyph_brush_ref
-                        .glyphs_custom_layout(&section, &layout)
-                        .map(|g| {
-                            let pos = g.position();
-                            let advance_width = g.unpositioned().h_metrics().advance_width;
-                            CachedGlyph {
-                                x: pos.x,
-                                y: -pos.y,
-                                advance_width,
-                            }
-                        }),
-                );
+                // `GlyphBrush::glyphs_custom_layout` does not return glyphs for invisible
+                // characters.
+                //
+                // <https://docs.rs/glyph_brush/0.6.2/glyph_brush/trait.GlyphCruncher.html
+                //  #tymethod.glyphs_custom_layout>
+                //
+                // For support, see:
+                //
+                // <https://github.com/alexheretic/glyph-brush/issues/80>
+                let mut nonempty_cached_glyphs = glyph_brush_ref
+                    .glyphs_custom_layout(&section, &layout)
+                    .map(|g| {
+                        let pos = g.position();
+                        let advance_width = g.unpositioned().h_metrics().advance_width;
+                        CachedGlyph {
+                            x: pos.x,
+                            y: -pos.y,
+                            advance_width,
+                        }
+                    });
+
+                let mut last_cached_glyph: Option<CachedGlyph> = None;
+                let all_glyphs = ui_text.text.chars().filter_map(|c| {
+                    if c.is_whitespace() {
+                        let (x, y) = if let Some(last_cached_glyph) = last_cached_glyph {
+                            let x = last_cached_glyph.x + last_cached_glyph.advance_width;
+                            let y = last_cached_glyph.y;
+                            (x, y)
+                        } else {
+                            (0.0, 0.0)
+                        };
+
+                        let advance_width =
+                            font_asset.glyph(c).scaled(scale).h_metrics().advance_width;
+
+                        let cached_glyph = CachedGlyph {
+                            x,
+                            y,
+                            advance_width,
+                        };
+                        last_cached_glyph = Some(cached_glyph);
```

Comment:
Maybe just do `Some(cached_glyph)` and remove the assignment and the line below.

Reply:
ah, the assignment is necessary, because `last_cached_glyph` lives beyond the closure, and is used for the whitespace calculation in subsequent iterations (same for next comment).